src/Decode.fs

```diff
@@ -154,239 +130,230 @@ module Decode =
-                (path, BadPrimitive("a string", token)) |> Error
+                Error (path, BadPrimitive("a string", token))
 
     let char : Decoder<char> =
         fun path value ->
             if Helpers.isString value then
                 let str = Helpers.asString value
                 if str.Length = 1 then
-                    Ok(str.[0])
+                    Ok str[0]
                 else
-                    (path, BadPrimitive("a single character string", value)) |> Error
+                    Error (path, BadPrimitive("a single-character string", value))
             else
-                (path, BadPrimitive("a char", value)) |> Error
+                Error (path, BadPrimitive("a char", value))
 
-    let guid : Decoder<System.Guid> =
+    let guid : Decoder<Guid> =
         fun path value ->
-            // Using Helpers.isString fails because Json.NET directly assigns Guid type
-            if Helpers.isGuid value then
-                value.Value<System.Guid>() |> Ok
-            elif Helpers.isString value then
-                match System.Guid.TryParse (Helpers.asString value) with
-                | true, x -> Ok x
-                | _ -> (path, BadPrimitive("a guid", value)) |> Error
-            else (path, BadPrimitive("a guid", value)) |> Error
+            if Helpers.isString value then
+                match value.TryGetGuid() with
+                | true, guid -> Ok guid
+                | _ -> Error (path, BadPrimitive("a guid", value))
+            else
+                Error (path, BadPrimitive("a guid", value))
 
     let unit : Decoder<unit> =
         fun path value ->
-            if Helpers.isNullValue value then
+            if Helpers.isNullValue value || Helpers.isUndefined value then
                 Ok ()
             else
-                (path, BadPrimitive("null", value)) |> Error
+                Error (path, BadPrimitive("null", value))
 
     let inline private integral
-                    (name : string)
-                    (tryParse : (string -> bool * 'T))
-                    (min : unit -> 'T)
-                    (max : unit -> 'T)
-                    (conv : float -> 'T) : Decoder< 'T > =
-
+        (name: string)
+        (rangeDescription: string)
+        (tryGet: JsonElement -> 'T option)
+        (tryParse : string -> bool * 'T) : Decoder<'T>
+        =
         fun path value ->
             if Helpers.isNumber value then
-                if Helpers.isIntegralValue value then
-                    let fValue = Helpers.asFloat value
-                    if (float(min())) <= fValue && fValue <= (float(max())) then
-                        Ok(conv fValue)
-                    else
-                        (path, BadPrimitiveExtra(name, value, "Value was either too large or too small for " + name)) |> Error
-                else
-                    (path, BadPrimitiveExtra(name, value, "Value is not an integral value")) |> Error
+                match tryGet value with
```

Comment:
this `integral` helper function was probably the most changed out of all of them because `System.Text.Json` only has a "number" type, it doesn't support "integer" as a type in the parsed AST. So, I adapted and changed the error reporting a bit